src/doorman/IdAllocator.hpp

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+class IdAllocator {
+    class ReservationObj {
+        IdAllocator &allocator_;
+        uint32_t id_;
+        friend IdAllocator;
+        ReservationObj(IdAllocator &allocator, uint32_t id) : allocator_(allocator), id_(id) {}
+
+    public:
+        ~ReservationObj() { allocator_.release(id_); }
+        ReservationObj(const ReservationObj &) = delete;
+        ReservationObj(ReservationObj &&) = delete;
+        ReservationObj &operator=(const ReservationObj &) = delete;
+        ReservationObj &operator=(ReservationObj &&) = delete;
+
+        [[nodiscard]] uint32_t id() const noexcept { return id_; }
+    };
+
+public:
+    using Reservation = std::shared_ptr<ReservationObj>;
+
+private:
+    uint32_t next_id_{};
+    std::vector<size_t> released_ids_;
```

Comment:
Should the type of the vector be uint32_t?

Reply:
Yes! Amazing spot. I changed from size_t to uin32_t late in the process. I will make an alias to keep the ID the same.